examples/exceptions2/policy/deny.rego

```diff
@@ -0,0 +1,29 @@
+package main
+
+exceptions = {"exception-name"}
+
+func_name_msg(name) = ret {
+	ret := sprintf("Resource Name '%s' contains dashes", [name])
+}
+
+deny_name[msg] {
+	input.resource[_][name]
+	contains(name, "-")
+	msg := func_name_msg(name)
+}
+
+deny_resource_type[msg] {
+	input.resource[type]
+	type == "invalid_type"
+	msg := sprintf("Resource Type '%s' is invalid", [type])
+}
+
+exclude_name[rules] {
+	input.resource[_][name]
+	exceptions[name]
+	rules := [func_name_msg(name)]
```

Comment:
Yeah, I know what you mean, I was thinking of potentially returning a complex object from the `deny` rules so that arbitrary data could be used to match an exclusion. Similar to what you suggest with the _magic_ inclusion up top.

This would mean changing the api for conftest though, so the impact would be quite a lot higher.

I believe the message abstraction would be sufficient for more complex cases too, as it would just be a string-generation function. I do agree though that it is then a requirement to enforce (socially) the use of the function to generate error messages. Something which currently isn't a requirement & is quite restful.

I like your idea of regex & substring matching, but I also know they have a tendency to be a little greedy when matching. Especially if they accidentally catch new violations.

I'll have a think about how to expand the deny-rule returns to allow complex objects so it's easier to make an exclusion based on the output of a deny rule (indirectly by comparing the output of the deny rule with the output of the exclusion rule).

This is effectively what the `func_name_message` function is doing, but in an ad-hoc manner. If this could be done in a more generic manner (ie. dynamic no. of key/value pairs.

I'll also try the example in the linked Issue with using k8s deployments as that may provide a different perspective.

Reply:
> I believe the message abstraction would be sufficient for more complex cases too, as it would just be a string-generation function. 

I'm not following this one - could you explain how it would work in the example of a rule that reports your name has N dashes instead of M?

> [regexp] have a tendency to be a little greedy when matching
100%

Reply:
> I'm not following this one - could you explain how it would work in the example of a rule that reports your name has N dashes instead of M?

Ah, I see what you're saying. Below is a small example of how to potentially work around it, but it does add more complexity.

```

msg_regex(name, count) = msg {
  msg := sprintf("resource %s has %d dashes where only 1 is allowed", [name, count])
}

count_dashes(str) {
  # logic to count no. dashes
}

deny_regex[msg] {
  count := count_dashes(name)
  msg := msg_regex(name, count)
}

exclude_regex[excludes] {
  name := count_dashes(name)
  excludes := [msg_regex(name, count)]
}
```

I have worked on a bit of a refactor, I'll push up the changes, with an example in a few hours.